### PR TITLE
Correct boost in script_score query and error on negative scores

### LINKED ISSUE #48465: Apply boost in script_score query to the whole query
Current `boost` parameter in script_score query is wrongly only applied to the subquery.
For example, this query returns `1` without applying `boost`:
```js
"query": {
    "script_score" : {
        "query" : {
            "match_all": {}
        },
        "script" : {
            "source" : "1"
        },
        "boost": 3
    }
}
```

While this query returns `3` applying boost to the subquery.
```js
"query": {
    "script_score" : {
        "query" : {
            "match_all": {}
        },
        "script" : {
            "source" : "_score"
        },
        "boost": 3
    }
}
```
This behaviour was partially copied from `function_score` query where `boost` is applied to both subquery and the whole query.  For example, this query returns `9` where boost is applied to the subquery and also the whole query (which is quite confusing).

```js
"query": {
    "function_score" : {
        "query" : {
            "match_all": {}
        },
        "script_score" : {
            "script" : {
                "source" : "_score"
            }
        },
        "boost": 3
    }
}
```

We need to correct `boost` behaviour in `script_score` to apply only the whole query, and not subquery.

### PR DESCRIPTION
**Pull request by mayya-sharipova**
Before boost in script_score query was wrongly applied only to the subquery.
This commit makes sure that the boost is applied to the whole score
that comes out of script.

Also provide error 400x error message on negative scores in script_score.

Closes #48465

### CODE REVIEW

File: docs/reference/query-dsl/script-score-query.asciidoc [hunk 1]
```diff
@@ -48,9 +48,12 @@ scores be positive or `0`.
 --
 
 `min_score`::
-(Optional, float) Documents with a <<relevance-scores,relevance score>> lower
+(Optional, float) Documents with a score lower
```
Comment (matriv) [inline docs/reference/query-dsl/script-score-query.asciidoc]: why did you remove the reference to the score docs?

File: server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java [hunk 2]
```diff
@@ -143,7 +148,12 @@ public Explanation explain(LeafReaderContext context, int doc) throws IOExceptio
                         explanation = Explanation.match(score, desc);
                     }
                 }
-                
+                if (boost != 1.0f) {
```
Comment (matriv) [inline server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java]: This seems to be executed also when `explanation == null`? Or I'm missing something?
Maybe it worths adding a test where the boost is ` != 1` and check that there is no explanation returned regarding the boost?

Also, why not return an explanation if the boost is `1`, since it's asked by the user?

File: modules/lang-painless/src/test/resources/rest-api-spec/test/painless/110_script_score_boost.yml [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+  - do:
+      search:
+        index: test_index
+        body:
+          query:
+            script_score:
+              query: {match_all: {boost: 10}}
+              script:
+                source: "doc['i'].value * _score"
+              boost: 10
```
Comment (matriv) [inline modules/lang-painless/src/test/resources/rest-api-spec/test/painless/110_script_score_boost.yml]: Maybe use a different score here, e.g.: `5` to make it more clear.

**Review by matriv — COMMENTED** [review]
Looks good, I'm completely inexperienced in the area but I left some comments, especially the one regarding explanation.

File: docs/reference/query-dsl/script-score-query.asciidoc (hunk 1 above)
Comment (mayya-sharipova) [reply docs/reference/query-dsl/script-score-query.asciidoc]: Before we were referencing relevance scores, while I think the goal of script_score query is to calculate custom scores through some scripts, not a traditional textual relevance score.

File: server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java (hunk 2 above)
Comment (mayya-sharipova) [reply server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java]: The way we handle `explanation` object is different from other queries, which may be confusing. In `script_score` query a user can provide his/her custom explanation. 

> This seems to be executed also when explanation == null?

Not exactly. We start with `explanation` that is user provided explanation for the script, and if it null, it is substituted by our standard explanation (line 143), so on in line 151 explanation is never null. 

> Maybe it worths checking for a test that when the boost is != 1 and explanation is false there is no explanation returned regarding the boost?

When is explanation is not asked (`explain = false` in a search request),  we will not even go to this method. This method is executed only when a user requests an explanation. 

> Also, why not return an explanation if the boost is 1, since it's asked by the user?

I was thinking since `boost` is an optional parameter,  when it is not provided by a user, there is no need to provide an explanation about it.

**Comment by mayya-sharipova** [comment]
@matriv Thanks for the review. I have tried to address your comments, please continue to review when you have time.

**Review by matriv — APPROVED** [review]
LGTM. Thanks a lot for responding to my questions and providing some more context!

File: server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java [hunk 4]
```diff
@@ -203,30 +213,33 @@ public int hashCode() {
...
 
         @Override
         public float score() throws IOException {
             int docId = docID();
             scoreScript.setDocument(docId);
-            float score = (float) scoreScript.execute(explanation);
+            float score = (float) scoreScript.execute(explanation) * boost;
             if (score == Float.NEGATIVE_INFINITY || Float.isNaN(score)) {
                 throw new ElasticsearchException(
                     "script_score query returned an invalid score [" + score + "] for doc [" + docId + "].");
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java]: I think it would be better if this error message returned the value produced by the script without the boost.

File: server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java [hunk 5]
```diff
@@ -143,7 +148,12 @@ public Explanation explain(LeafReaderContext context, int doc) throws IOExceptio
                         explanation = Explanation.match(score, desc);
                     }
                 }
-                
+                if (boost != 1.0f) {
+                    List<Explanation> subs = new ArrayList<>();
+                    subs.addAll(Arrays.asList(explanation.getDetails()));
+                    subs.add(Explanation.match(boost, "boost"));
+                    explanation = Explanation.match(explanation.getValue(), explanation.getDescription(), subs);
+                }
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java]: I'd suggest wrapping the explanation instead of modifying it in-place: create the scorer with boost=1f a couple lines above, and then here:

```java
if (boost != 1f) {
    explanation = Explanation.match(boost * explanation.getValue().floatValue(), "Boosted score, product of:",
        Explanation.match(boost, "boost"),
        explanation);
}
```

**Review by jpountz — APPROVED** [review]
I left a minor suggestion, otherwise LGTM.

**Comment by mayya-sharipova** [comment]
@jpountz Thanks for the feedback, I have addressed it.
In the [last commit](https://github.com/elastic/elasticsearch/pull/52478/commits/96db563f4c5eb0e3c9e694976db0c2b65ed2dd9b#diff-343986e8723b9c666edbb9c2d45dd2f0R233) , I have also corrected the exception to make sure a 400 error is returned when a score is negative.

File: server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java [hunk 6]
```diff
@@ -221,12 +225,13 @@ public float score() throws IOException {
             int docId = docID();
             scoreScript.setDocument(docId);
             float score = (float) scoreScript.execute(explanation);
-            if (score == Float.NEGATIVE_INFINITY || Float.isNaN(score)) {
-                throw new ElasticsearchException(
-                    "script_score query returned an invalid score [" + score + "] for doc [" + docId + "].");
+            if (score < 0f || Float.isNaN(score)) {
```
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java]: It looks like this PR also fixed a bug in `script_score` queries where we allowed negative scores. I think we should add a note to the breaking changes docs and also update the PR description to make it clear we included this change.

Comment (mayya-sharipova) [reply server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java]: @jtibshirani Thanks.  Even before without this change a user would get an error if their `script_score` query produced a negative score. They would just get it from a different place, I think from the Lucene [here](https://github.com/apache/lucene-solr/blob/master/lucene/core/src/java/org/apache/lucene/search/TopScoreDocCollector.java#L79)

So the only thing changed from a user perspective is an error message and error status code (before was 500, not 400x).  Do you think it warrants a breaking change notice?
+1 for include this in the PR description

Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java]: I tried this out using an Elasticsearch 7.6 build, and didn't receive an error:

```
PUT my_index/_doc/1?refresh
{
  "field": "value"
}

GET my_index/_search
{
  "query": {
    "script_score": {
      "query": {
        "match_all": {}
      },
      "script": {
        "source": "-1000"
      }
    }
  }
}
```

The line you linked to is an assert, so perhaps these Lucene checks didn't always catch the issue in non-test environments.

Comment (mayya-sharipova) [reply server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java]: @jtibshirani Thanks for uncovering this.  I understood what happened:
- Before 7.5,  `script_score` query was using [ScriptScoreFunction](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreFunction.java#L83) that was returning 400 error with a negative score.  
- From 7.5, we have changed it to not use `ScriptScoreFunction` but forgot to add a condition for a negative score.  But  [TopScoreDocCollector assertion](https://github.com/apache/lucene-solr/blob/master/lucene/core/src/java/org/apache/lucene/search/TopScoreDocCollector.java#L79) is tripped, causing fatal error in the dev mode. But I guess we silence these assertions in a production mode  as we don't see any visible errors or error log messages.

So, negative scores were wrongly allowed only in 7.5-7.6 versions, so to me it doesn't look like  a breaking change. WDYT?

Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java]: Adding an explanation to the release notes makes sense to me. I agree it shouldn't be presented as a typical 'breaking change', it is more like a regression that we fixed. Perhaps we could add a unit test along with the release notes update, to prevent a future regression?
